[Fix-17004] Fix process termination logic in cancelApplication method

davidzollo opened the pull request:
- update the ProcessUtils.kill method to ensure it correctly handles process tree termination and direct process destruction.
- update the AbstractCommandExecutor.cancelApplication method to ensure it correctly handles process termination, including logging and error handling.

close https://github.com/apache/dolphinscheduler/issues/17004

davidzollo committed:
[Fix-17004] Fix process termination logic in AbstractCommandExecutor.cancelApplication

davidzollo committed:
Merge branch 'dev' into dev

davidzollo committed:
Merge branch 'dev' into dev

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/AbstractCommandExecutor.java:
```diff
@@ -216,13 +216,30 @@ public void cancelApplication() throws InterruptedException {
+                return;
+            }
+
+            // If killing process tree fails, try to destroy the process directly
+            log.info("Failed to kill process tree, trying to destroy process directly");
+            process.destroy();
+            if (!process.waitFor(5, TimeUnit.SECONDS)) {
+                log.info("Process did not exit after destroy, forcing termination");
+                process.destroyForcibly();
+            }
+            log.info("Successfully killed process tree for task: {}, pid: {}",
+                    taskRequest.getTaskAppId(), taskRequest.getProcessId());
+
+        } catch (Exception e) {
+            log.error("Error while killing process, pid: {}", taskRequest.getProcessId(), e);
+            // Try destroyForcibly as last resort
             process.destroyForcibly();
```
This is not needed.

davidzollo replied:
yes, good suggesion.

davidzollo replied:
I have been removed these logic,thx

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtils.java:
```diff
@@ -85,31 +85,105 @@ private ProcessUtils() {
             int processId = request.getProcessId();
             if (processId == 0) {
-                log.error("Task instance kill failed, processId is not exist");
+                log.error("Task instance kill failed, processId is not available");
```
We should return true here, the logic is equals with `pidArray` is empty.

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtils.java:
```diff
@@ -85,31 +85,105 @@ private ProcessUtils() {
+            }
+
+            // 1. Try to terminate gracefully (SIGINT)
+            boolean gracefulKillSuccess = sendKillSignal("SIGINT", pids, request.getTenantCode(), 2000);
```
Don't add wait time, this will make RPC failed.

davidzollo replied:
I have removed this wait time

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/ProcessUtils.java:
```diff
@@ -85,31 +85,105 @@ private ProcessUtils() {
 
-            OSUtils.exeCmd(cmd);
-            log.info("Success kill task instance, processId: {}", request.getProcessId());
+            // 4. Finally, check if the process is still alive
+            for (String pid : pidArray) {
+                if (isProcessAlive(pid)) {
+                    log.error("Failed to kill process {}, even after multiple attempts", pid);
+                    return false;
+                }
+            }
+
+            log.info("Success: process {} is no longer running", processId);
```
Is this needed to do a check here, if the process has set graceful exist, this check will false, but this doesn't mean we kill failed.

davidzollo replied:
You're right, it's no need to check here.

ruanwenjun reviewed (commented):
Is this PR generating by AI? This could potentially lead to high code duplicate with other project.

davidzollo commented:
> Is this PR generating by AI? This could potentially lead to high code duplicate with other project.

At first, I use cursor generate PR code, but there're many errors it generated, so I have to update a lot to solve this issue.

davidzollo committed:
Merge branch 'dev' into dev

davidzollo committed:
Merge remote-tracking branch 'upstream/dev' into dev

davidzollo committed:
Merge remote-tracking branch 'origin/dev' into dev

davidzollo committed:
[Fix-17004] Fix process termination logic in AbstractCommandExecutor.cancelApplication

davidzollo commented:
@ruanwenjun Thanks for the review, I have updated the PR, please help review it when you're free, thx

ruanwenjun commented:
> @ruanwenjun Thanks for the review, I have updated the PR, please help review it when you're free, thx

It's better to remove the extra code in https://github.com/apache/dolphinscheduler/pull/17005#discussion_r1957241984

davidzollo committed:
[Fix-17004] Fix process termination logic in AbstractCommandExecutor.cancelApplication

davidzollo commented:
> > @ruanwenjun Thanks for the review, I have updated the PR, please help review it when you're free, thx
> 
> It's better to remove the extra code in [#17005 (comment)](https://github.com/apache/dolphinscheduler/pull/17005#discussion_r1957241984)

Thanks for the suggestion, I have removed this logic, please help review it when you're free, thx

ruanwenjun reviewed (approved):
LGTM

SbloodyS reviewed (approved):
+1
